src/analysis/normed_space/star/multiplier.lean

```diff
@@ -0,0 +1,414 @@
+@[simp] lemma smul_right (k : 𝕜) (a : 𝓜(𝕜, A)) : (k • a).right = k • a.right := rfl
+
+end scalars
+
+@[simp] lemma add_left (a b : 𝓜(𝕜, A)) : (a + b).left = a.left + b.left := rfl
+@[simp] lemma add_right (a b : 𝓜(𝕜, A)) : (a + b).right = a.right + b.right := rfl
+@[simp] lemma zero_left : (0 : 𝓜(𝕜, A)).left = 0 := rfl
+@[simp] lemma zero_right : (0 : 𝓜(𝕜, A)).right = 0 := rfl
+@[simp] lemma neg_left (a : 𝓜(𝕜, A)) : (-a).left = -a.left := rfl
+@[simp] lemma neg_right (a : 𝓜(𝕜, A)) : (-a).right = -a.right := rfl
+@[simp] lemma sub_left (a b : 𝓜(𝕜, A)) : (a - b).left = a.left - b.left := rfl
+@[simp] lemma sub_right (a b : 𝓜(𝕜, A)) : (a - b).right = a.right - b.right := rfl
+
+/-- The module structure is inherited as the pullback under the injective map
+`double_centralizer.prod_mk : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)` -/
+instance : add_comm_group 𝓜(𝕜, A) :=
+function.injective.add_comm_group (prod_mk 𝕜 A) injective_prod_mk rfl (λ x y, rfl) (λ x, rfl)
+  (λ x y, rfl) (λ x n, rfl) (λ x n, rfl)
+
+/-- The canonical map `double_centralizer.prod_mk` as an additive group homomorphism. -/
+def add_group_hom_prod_mk : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A) :=
+{ to_fun := prod_mk 𝕜 A,
+  map_zero' := rfl,
+  map_add' := λ x y, rfl }
+
+/-- The module structure is inherited as the pullback under the additive group monomoprhism
+`double_centralizer.prod_mk : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A)` -/
+instance : module 𝕜 𝓜(𝕜, A) :=
+function.injective.module 𝕜 add_group_hom_prod_mk injective_prod_mk (λ x y, rfl)
+
+/-- The normed group structure is inherited as the pullback under the additive group monomoprhism
+`double_centralizer.prod_mk : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A)` -/
+instance : normed_add_comm_group 𝓜(𝕜, A) :=
+normed_add_comm_group.induced _ _ (add_group_hom_prod_mk : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A))
+  injective_prod_mk
+
+@[simp] lemma norm_eq (a : 𝓜(𝕜, A)) : ∥a∥ = max (∥a.left∥) (∥a.right∥) := rfl
+
+instance : normed_space 𝕜 𝓜(𝕜, A) :=
+{ norm_smul_le := λ k a, show max (∥k • a.left∥) (∥k • a.right∥) ≤ ∥k∥ * max (∥a.left∥) (∥a.right∥),
+    by simp only [mul_max_of_nonneg _ _ (norm_nonneg k), norm_smul],
+  .. double_centralizer.module }
+
+lemma uniform_embedding_prod_mk : uniform_embedding (prod_mk 𝕜 A) :=
+uniform_embedding_comap injective_prod_mk
+
+instance [complete_space A] : complete_space 𝓜(𝕜, A) :=
+begin
+  rw complete_space_iff_is_complete_range uniform_embedding_prod_mk.to_uniform_inducing,
+  apply is_closed.is_complete,
+  simp only [range_prod_mk, set.set_of_forall],
+  refine is_closed_Inter (λ x, is_closed_Inter $ λ y, is_closed_eq _ _),
+  { exact ((continuous_mul_right y).comp (continuous_linear_map.apply 𝕜 A x).continuous).comp
+      continuous_snd },
+  { exact ((continuous_mul_left x).comp (continuous_linear_map.apply 𝕜 A y).continuous).comp
+      continuous_fst }
+end
+
+/-!
+### Multiplicative structure
+-/
+
+instance : ring 𝓜(𝕜, A) :=
```

Comment:
This ring structure is obtained by pulling back the one from `module.End 𝕜 A × module.End 𝕜 A` under `prod_mk`. Can you define it that way? It should make everything shorter.

Reply:
I have now implemented the ring structure as the pullback under the natural map `𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ`. I'm not sure whether it's actually shorter (because you still have to define `has_pow`, `has_nat_cast` and `has_int_cast` instances, but it's at least a bit more elegant.